File: src/python/graph/bfs/tests/traversal_tests.py
import sys
import pytest

from traversal import traverse


graph = {
    # first subgraph
    "ROOT": ["B", "C", "D"],
    "B": ["E"],
    "C": ["F", "G"],
    "F": ["H", "I"],

    # second subgraph
    "J": ["K", "L", "M"],
    "L": ["N", "O"]
}


def test_successful_traversal():
    first_subgraph = traverse(graph, root="ROOT")
    assert first_subgraph == ["ROOT", "B", "C", "D", "E", "F", "G", "H", "I"]

    second_subgraph = traverse(graph, root="J")
    assert second_subgraph == ["J", "K", "L", "M", "N", "O"]


def test_empty_path_if_invalid_root():
    result = traverse(graph, root="invalid_vertex")
    assert result == None


if __name__ == "__main__":
    sys.exit(pytest.main([__file__]))

File: src/swift/graph/bfs/tests/TraversalTests.swift
import XCTest


class TraversalTests: XCTestCase {

    let graph = [
        // first subgraph
        "A": ["B", "C", "D"],
        "B": ["E"],
        "C": ["F", "G"],
        "F": ["H", "I"],

        // second subgraph
        "J": ["K", "L", "M"],
        "L": ["N", "O"],       
    ]


    func test_successful_traversal() {
        let firstSubgraph = traverse(in: graph, withRoot: "A")
        XCTAssertEqual(firstSubgraph, ["A", "B", "C", "D", "E", "F", "G", "H", "I"])

        let secondSubgraph = traverse(in: graph, withRoot: "J")
        XCTAssertEqual(secondSubgraph, ["J", "K", "L", "M", "N", "O"])
    }

    func test_nil_if_invalid_root() {
        let result = traverse(in: graph, withRoot: "InvalidVertex")
        XCTAssertNil(result)
    }


    static var allTests = [
        ("test_successful_traversal", test_successful_traversal),
        ("test_nil_if_invalid_root", test_nil_if_invalid_root),
    ]
}


func testMain() {
    #if os(Linux)

    XCTMain([
        testCase(TraversalTests.allTests)
    ])
    #endif
}